MoltenVK/MoltenVK/Commands/MVKMTLBufferAllocation.mm

```diff
@@ -89,11 +99,19 @@
     // Convert max length to the next power-of-two exponent to use as a lookup
     NSUInteger p2Exp = mvkPowerOfTwoExponent(length);
 	MVKMTLBufferAllocationPool* pRP = _regionPools[p2Exp];
-	const MVKMTLBufferAllocation* region = _isThreadSafe ? pRP->acquireObjectSafely() : pRP->acquireObject();
+    if (_isThreadSafe)
+        pRP->_lock.lock();
+
+    MVKMTLBufferAllocation* region = pRP->acquireObject();
 	if (region) {
+        if (!(*region->_lockCount)++)
+            mlock(region->_mtlBuffer.contents, region->_mtlBuffer.length);
 		[region->_mtlBuffer setPurgeableState: MTLPurgeableStateVolatile];
 	}
```

Comment:
I touched on that in the comment above, in addition, locking/unlocking parts of a `MTLBuffer` would work just fine (and wouldn't require counting) if the parts were page-aligned. Since they aren't, we may end up unlocking more than we wanted to. In princinple the whole `MTLBuffer` can be filled with allocations that cross page boundaries, so you'd need to track which individual pages are good to `munlock`, and that would be overkill as the temporary `MTLBuffer`s aren't that large anyway.
One assumption I didn't verify is that `MTLBuffer`s are always given their own, independent memory mappings.

Reply:
>I touched on that in the comment above

Yes...and I'm still working to understand the restrictions. Are you saying that unlocking an individual allocation that spanned two pages would unlock both pages, even if other `mlocks()` have been placed elsewhere in the those pages? 

That doesn't seem like a functional design for `mlock()/munlock()`, because locking memory doesn't mean anything if it can arbitrarily be unlocked as a side-effect of nearby unlocks.

And extending that problem to your design...what if two `MTLBuffers` overlap a single memory page? Would unlocking one `MTLBuffer` unlock the other?

Reply:
> Yes...and I'm still working to understand the restrictions. Are you saying that unlocking an individual allocation that spanned two pages would unlock both pages, even if other `mlocks()` have been placed elsewhere in the those pages?
> 
> That doesn't seem like a reasonable functional design for `mlock()/munlock()`, because locking memory doesn't mean anything if it can arbitrarily be unlocked as a side-effect of nearby unlocks.

For `mlock/munlock` the restriction is that they (like other parts of the VM machinery) operate on whole pages only. So if we have a few small temporary allocations in a single `MTLBuffer` next to each other, fitting in a single page, there's no way to single out only one of these regions - `mlock/munlock` will round up to whole pages anyway.
Now regarding other locks: it is usually the case that they do not stack, i.e. if you lock some pages twice, they are unlocked after a single `munlock()`. So indeed, you need to be page-granularity aware if you don't want to unlock something by accident.
Now I say usually, and to be fair I just noticed that now, on macOS `man mlock` says:
```
BUGS
     Unlike The Sun implementation, multiple mlock calls on the same address
     range require the corresponding number of munlock calls to actually
     unlock the pages, i.e.  mlock nests.  This should be considered a consequence
     of the implementation and not a feature.
```
So we probably should not rely on that behavior.
 
> And extending that problem to your design...what if two `MTLBuffers` overlap a single memory page? Would unlocking one `MTLBuffer` unlock the other?

That's why I mentioned it as an assumption - I would be quite surprised if that could happen. For one, because it would be an unexpected foot gun for this exact reason. I expect each `MTLBuffer` to have its own mapping in the VM system - and these have page granularity as well. I guess that it's possible for two `MTLBuffer`s to share a single physical memory page, that should not be an issue though, as long as they get separate VM mappings.

Reply:
> This should be considered a consequence of the implementation and not a feature.

"It's not a feature, it's a bug!"...I love it. 🤣🤦🏻‍♂️

Reply:
Hmmm...we might have a way to avoid the locking issue entirely because Metal looks to be handling this for us.

At 39:00 in [this tutorial](https://developer.apple.com/videos/play/wwdc2019/606/), Apple explains that `setPurgeableState:` should be toggled between `MTLPurgeableStateVolatile` and `MTLPurgeableStateNonVolatile` as the `MTLBuffer` enters and exists the resource pool, respectively. 

With that in mind..rather than using `mlock()` and `munlock()` at all, we should be able to just toggle `setPurgeableState:` appropriately as we track that all regions in the `MTLBuffer` have been returned to the allocation pool. In other words, the `MTLBuffer` is marked volatile only while none of it is in use outside the allocation pool.

Reply:
> track that all regions in the `MTLBuffer`

 struct LockedBuffer { id&lt;MTLBuffer&gt; mtlBuffer; uint64_t lockCount; };    MVKSmallVector&lt;LockedBuffer, 64&gt; _lockedBuffers;

Maybe rename `LockedBuffer`, `lockCount`, and `_lockedBuffers` to remove the idea of "lock" (eg. something like `MTLBufferTracker`, `allocatedCount`, and `_mtlBuffers`, respectively), and work with that.

Reply:
>  `setPurgeableState:` should be toggled between `MTLPurgeableStateVolatile` and `MTLPurgeableStateNonVolatile` as the `MTLBuffer` enters and exists the resource pool
Yeah, that looks like the preferred way of doing it. Thanks for the research - I pushed the updated commit.